packages/volto/src/components/manage/UniversalLink/UniversalLink.tsx

```diff
@@ -0,0 +1,205 @@
+import React from 'react';
+import { HashLink as Link } from 'react-router-hash-link';
+import { useSelector } from 'react-redux';
+import {
+  flattenToAppURL,
+  isInternalURL,
+  URLUtils,
+} from '@plone/volto/helpers/Url/Url';
+import config from '@plone/volto/registry';
+import cx from 'classnames';
+
+type BaseProps = {
+  openLinkInNewTab?: boolean;
+  download?: boolean;
+  children: React.ReactNode;
+  className?: string;
+  title?: string;
+  smooth?: boolean;
+  onClick?: (e: React.MouseEvent) => void;
+  onKeyDown?: (e: React.KeyboardEvent) => void;
+};
+
+type HrefOnly = {
+  href: string;
+  item?: never;
+} & BaseProps;
+
+type ItemOnly = {
+  href?: never;
+  item: {
+    '@id': string;
+    remoteUrl?: string;
+    getRemoteUrl?: string;
+    '@type'?: string;
+  };
+} & BaseProps;
+
+export type UniversalLinkProps = HrefOnly | ItemOnly;
+export interface AppState {
+  content: {
+    data?: {
+      '@components'?: {
+        translations?: {
+          items?: { language: string; '@id': string }[];
+        };
+      };
+    };
+  };
+  navroot: {
+    data: {
+      navroot: {
+        id: string;
+      };
+    };
+  };
+  intl: {
+    locale: string;
+  };
+  userSession: {
+    token: string;
+  };
+}
+
+export const __test = {
+  renderCounter: () => {},
+};
+
+const UniversalLink = React.memo(function UniversalLink(
+  props: UniversalLinkProps,
+) {
+  const {
+    openLinkInNewTab,
+    download,
+    children,
+    className,
+    title,
+    smooth,
+    onClick,
+    onKeyDown,
+    item,
+    ...rest
+  } = props;
+  __test.renderCounter();
+
+  const token = useSelector<AppState, string>(
+    (state) => state.userSession?.token,
+  );
+
+  function getUrl(props: UniversalLinkProps, token: string): string {
```

Comment:
I'm fine with turning the previous code into a function, if we go this way I would extract the function from the component completely so it's also easily testable. This is not mandatory but would be nice imho.

Reply:
Done. (not tests, but I separated it).